File: TJAVA (1º trabalho)/código/src/Candidato.java
package src;

import java.lang.String;

public class Candidato implements Comparable<Candidato>{
    private int numero; //Número do candidato na urna;
    private int votosNominais; //Total de votos nominais do candidato;
    private String situacao; //Situação do candidato: Eleito / Não Eleito / Suplente;
    private String nome; //Nome completo do candidato;
    private String nomeUrna; //Nome do candidato como mostrado na urna eletrônica;
    private String sexo; //Sexo do candidato: F / M;
    private String dataNasc; //Data de nascimento do candidato no formato dd/mm/aaaa;
    private String destinoVoto; //Válido / Anulado / Anulado sub judice; 
    private int numeroPartido; //Número do partido do candidato;
    private int pos = 0;
    private int flagCompare = 0;
    
    Candidato() {}

    Candidato(int numero, int votosnominais, String situacao,String nome, String nomeUrna, String sexo,String dataNasc, String destinoVoto, int numeroPartido){
        this.numero = numero;
        this.votosNominais = votosnominais;
        this.situacao = situacao;
        this.nome = nome;
        this.nomeUrna = nomeUrna;
        this.sexo = sexo;
        this.dataNasc = dataNasc;
        this.destinoVoto = destinoVoto;
        this.numeroPartido = numeroPartido;
    }

    //GETs e SETs
    
    public void setPos(int pos){
        this.pos = pos;
    }
    public int getPos(){
        return this.pos;
    }

    public void setFlagCompare(int flagCompare){
        this.flagCompare = flagCompare;
    }
    public int getFlagCompare(){
        return this.flagCompare++;
    }

    public void setNumero (int numero){
        this.numero = numero;
    }

    public void setVotosNominais (int votosNominais){
        this.votosNominais = votosNominais;
    }
    
    public void setSituacao (String situacao){
        this.situacao = situacao;
    }
    
    public void setNome (String nome){
        this.nome = nome;
    }

    public void setNomeUrna (String nomeUrna){
        this.nomeUrna = nomeUrna;
    }

    public void setSexo (String sexo){
        this.sexo = sexo;
    }

    public void setDataNasc (String dataNasc){
        this.dataNasc = dataNasc;
    }

    public void setDestinoVoto (String destinoVoto){
        this.destinoVoto = destinoVoto;
    }

    public void setNumeroPartido (int numeroPartido){
        this.numeroPartido = numeroPartido; 
    }

    public int getNumero(){
        return this.numero;
    }
    
    public int getVotosNominais(){
        return this.votosNominais;
    }
    
    public String getSituacao(){
        return this.situacao;
    }
    
    public String getNome(){
        return this.nome;
    }

    public String getNomeUrna(){
        return this.nomeUrna;
    }

    public String getSexo(){
        return this.sexo;
    }

    public String getDataNasc(){
        return this.dataNasc;
    }
    
    public String getDestinoVoto(){
        return this.destinoVoto;
    }

    public int getNumeroPartido(){
        return this.numeroPartido;
    }

    public int retornaIdadeCandidato(String dataEleicao){
        String[] dataEleicaoSeparada = dataEleicao.split("/");
        String[] dataNascSeparada = this.dataNasc.split("/");

        int idade = 0;
        int diaEleicao = 0, diaNasc = 0;
        int MesEleicao = 0, MesNasc = 0;
        int AnoEleicao = 0, AnoNasc = 0;
        try{
        diaEleicao = Integer.parseInt(dataEleicaoSeparada[0]);
        MesEleicao = Integer.parseInt(dataEleicaoSeparada[1]);
        AnoEleicao = Integer.parseInt(dataEleicaoSeparada[2]);
        diaNasc = Integer.parseInt(dataNascSeparada[0]);
        MesNasc = Integer.parseInt(dataNascSeparada[1]);
        AnoNasc = Integer.parseInt(dataNascSeparada[2]);
        }
        catch(NumberFormatException e)
        {
        //Caso nao seja inteiro executa o que estiver neste bloco
        }

        if( MesNasc < MesEleicao){
            idade = AnoEleicao - AnoNasc;
        }else if(MesNasc == MesEleicao){
            if(diaNasc <= diaEleicao){
                idade = AnoEleicao - AnoNasc;
            }else{
                idade = (AnoEleicao - AnoNasc) - 1;
            }
        }else if(MesNasc > MesEleicao){
            idade = (AnoEleicao - AnoNasc) - 1;
        }
        return idade;
    }
    
    @Override
    public int compareTo(Candidato candidatos){
        
        if(this.votosNominais < candidatos.getVotosNominais()){
            return 1;
        }
        else if(this.votosNominais > candidatos.getVotosNominais()){
            return -1;
        }   
        if(candidatos.flagCompare == 1){
            if(this.numeroPartido < candidatos.numeroPartido){
                return -1;
            }else{
                return 1;
            }
        }else{
            /*
            if(this.retornaIdadeCandidato("15/11/2020") < candidatos.retornaIdadeCandidato("15/11/2020")){
                return 1;
            }else{
                return -1;
            }
        }
        */

        if(Integer.parseInt(this.getDataNasc().substring(6,9)) > Integer.parseInt(candidatos.getDataNasc().substring(6,9))){
            return 1;
        }else if(Integer.parseInt(this.getDataNasc().substring(6,9)) == Integer.parseInt(candidatos.getDataNasc().substring(6,9))){
            if(Integer.parseInt(this.getDataNasc().substring(3,4)) > Integer.parseInt(candidatos.getDataNasc().substring(3,4))){
                return 1;
            }else if(Integer.parseInt(this.getDataNasc().substring(3,4)) == Integer.parseInt(candidatos.getDataNasc().substring(3,4))){
                if((Integer.parseInt(this.getDataNasc().substring(0,1)) > Integer.parseInt(candidatos.getDataNasc().substring(0,1)))){
                    return 1;
                }else{
                    return -1;
                }
            }else{
                return -1;
            }
        }else{
            return -1;
        }
    }
}

    public String toString(String nomePartido){
        return String.format("%s / %s (%s, %d votos)", this.nome, this.nomeUrna,nomePartido, this.votosNominais);
    }

}

File: TC++ (2º trabalho)/código/Candidato.cpp
#include "Candidato.h"
#include <cstdio>
#include <string>
#include <vector>
#include <iostream>
#include <fstream>
#include <sstream>

// Construtores

Candidato::Candidato() {};

Candidato::Candidato(int numero, int votosNominais, const string &situacao,const string& nome, const string& nomeUrna,const string& sexo,const string &dataNasc,const string&destinoVoto, int numeroPartido, int pos) {
    this->numero = numero;
    this->votosNominais = votosNominais;
    this->situacao = situacao;
    this->nome = nome;
    this->nomeUrna = nomeUrna;
    this->sexo = sexo;
    this->dataNasc = dataNasc;
    this->destinoVoto = destinoVoto;
    this->numeroPartido = numeroPartido;
    this->pos = pos;
    this->flagCompare = 0;
};

// GETs & SETs

void Candidato::setNumero(int numero){
    this->numero = numero;
}
const int Candidato::getNumero() const{
    return this->numero;
}

void Candidato::setVotosNominais(int votosNominais){
    this->votosNominais = votosNominais;
}
const int Candidato::getVotosNominais() const{
    return this->votosNominais;
}   

void Candidato::setSituacao(const string& situacao){
    this->situacao = situacao;
}
const string& Candidato::getSituacao() const{
    return this->situacao;
}   

void Candidato::setNome(const string& nome){
    this->nome = nome;
}
const string& Candidato::getNome() const{
    return this->nome;
}

void Candidato::setNomeUrna(const string& nome){
    this->nomeUrna = nome;
}
const string& Candidato::getNomeUrna() const{
    return this->nomeUrna;
}

void Candidato::setSexo(const string& sexo){
    this->sexo = sexo;
}
const string& Candidato::getSexo() const{
    return this->sexo;
}

void Candidato::setDataNasc(const string& data){
    this->dataNasc = data;
}
const string& Candidato::getDataNasc() const{
    return this->dataNasc;
}

void Candidato::setDestinoVoto(const string& destinoVoto){
    this->destinoVoto = destinoVoto;
}
const string& Candidato::getDestinoVoto() const{
    return this->destinoVoto;
}

void Candidato::setNumeroPartido(int numeroPartido){
    this->numeroPartido = numeroPartido;
}
const int Candidato::getNumeroPartido() const{
    return this->numeroPartido;
}

void Candidato::setPos(int pos){
    this->pos = pos;
}
const int Candidato::getPos() const{
    return this->pos;
}

void Candidato::setFlagCompare(int flagCompare){
    this->flagCompare = flagCompare;
}
const int Candidato::getFlagCompare() const{
    return this->flagCompare;
}

// Métodos

const bool Candidato::operator==(const Candidato& c1) const{
    if(this->getNome() == c1.getNome()){
        return true;
    }else {
        return false;
    }
}

const int Candidato::retornaIdadeCandidato(const string& dataEleicao) const{
    vector<string> dataEleicaoSeparada;

    // variável de trabalho
    string token;

    //Lendo linha por linha.
    // stream de strings de input inicializado com a string a ser separada
    istringstream tokenizer { dataEleicao };

    //separando a linha lida pelas vírgulas.
    while (getline(tokenizer, token, '/'))
        dataEleicaoSeparada.push_back(token);


    vector<string> dataNascSeparada;
    string dataNasc = this->getDataNasc();
    // variável de trabalho
    string token2;

    //Lendo linha por linha.
    // stream de strings de input inicializado com a string a ser separada
    istringstream tokenizer2 { dataNasc };

    //separando a linha lida pelas vírgulas.
    while (getline(tokenizer2, token2, '/'))
        dataNascSeparada.push_back(token2);

    int idade = 0;
    int diaEleicao = 0, diaNasc = 0;
    int MesEleicao = 0, MesNasc = 0;
    int AnoEleicao = 0, AnoNasc = 0;
    diaEleicao = stoi(dataEleicaoSeparada[0]);
    MesEleicao = stoi(dataEleicaoSeparada[1]);
    AnoEleicao = stoi(dataEleicaoSeparada[2]);
    diaNasc = stoi(dataNascSeparada[0]);
    MesNasc = stoi(dataNascSeparada[1]);
    AnoNasc = stoi(dataNascSeparada[2]);
    
    if( MesNasc < MesEleicao){
        idade = AnoEleicao - AnoNasc;
    }else if(MesNasc == MesEleicao){
        if(diaNasc <= diaEleicao){
            idade = AnoEleicao - AnoNasc;
        }else{
            idade = (AnoEleicao - AnoNasc) - 1;
        }
    }else if(MesNasc > MesEleicao){
        idade = (AnoEleicao - AnoNasc) - 1;
    }
    return idade;
}

bool Candidato::compareVotosNominais(const Candidato& c1,const Candidato& c2) {
    if(c1.getVotosNominais() > c2.getVotosNominais()){
        return true;
    }else if(c1.getVotosNominais() < c2.getVotosNominais()){
        return false;
    }else{
        if(c2.getFlagCompare() == 1){
            if(c1.getNumeroPartido() < c2.getNumeroPartido()){
                return true;
            }else{
                return false;
            }
        }else{
            if(stoi(c1.getDataNasc().substr(6,4)) > stoi(c2.getDataNasc().substr(6,4))){
                return false;
            }
            else if(stoi(c1.getDataNasc().substr(6,4)) == stoi(c2.getDataNasc().substr(6,4))){
                if(stoi(c1.getDataNasc().substr(3,2)) > stoi(c2.getDataNasc().substr(3,2))){
                    return false;
                }else if(stoi(c1.getDataNasc().substr(3,2)) == stoi(c2.getDataNasc().substr(3,2))){
                    if((stoi(c2.getDataNasc().substr(0,2)) > stoi(c2.getDataNasc().substr(0,2)))){
                        return false;
                    }else{
                        return true;
                    }
                }else{
                    return true;
                }
            }else{
                return true;
            }
        }
    }
}